Refactored Authorization into an interface

zealws opened the pull request:
This PR adds new authorization code to the certauth middleware with a few improvements:

- Authorization has been generalized into an interface: `AuthorizationChecker`
  Multiple `AuthorizationChecker`s can be provided in the `certauth.Options` struct and they must all pass for a request to be allowed.
  This obsoletes the `AllowedOUs` and `AllowedCNs` fields on the `certauth.Options` struct.
- `AuthorizationChecker` may return arbitrary key/value pairs which are merged into the request's `context` (if Authorization is allowed) before running the application's handler.

For now this is only the interface (and backward compatibility for `AllowedOUs` and `AllowedCNs`). The Pantheon site authorization code is going to go into a separate PR which I am still working on.

EDIT: I updated this to reflect the latest state of the PR as the original goal of this PR has changed since I first wrote this summary.

vtopc commented on authorization.go:
```diff
@@ -0,0 +1,204 @@
+// Also validates that the site ID is a valid UUID.
+// Returns (site, environment, nil) if the clientCN is valid.
+// Returns ("", "", err) if an error occurs.
+func ParseSiteEnvFromCN(clientCN string) (string, string, error) {
```
I would make named return instead of docstring:
```suggestion
func ParseSiteEnvFromCN(clientCN string) (site string, environment string, err error) {
```
https://golang.org/doc/effective_go#named-results

Or return some struct
```
type ParseSiteEnvFromCNResp struct {
    Site string
    Environment string
}
```

winmillwill replied:
-1 on named return. Feature shouldn't exist if less really is exponentially more, as language designers claim.
+1 on struct. Maybe just one func that takes the whole subject and returns a struct holding all the subject info AND the stuff parsed out.

zealws replied:
> I would make named return instead of docstring:

Lol. I originally wrote it this way using a named return cause it's much clearer what each thing is returning.

But on previous PRs, several people got angry at me for using named returns like this, so I restructured this to avoid comments like that.

I'd much rather use a named return here as I think it's clearer what is returned by the function, and the function is simple enough I'm not too worried about the issues around named returns.

But honestly, the way it is now, it's obvious what's returned by the comment on line 161 and 162. And by the return statements throughout the function.

Wrapping the return in the struct might be helpful if more information gets embedded in the CN later that we need to parse out, but I don't think it makes this function more readable as-is.

vtopc replied:
Sorry, I'm a newbie here, just wanted to mention [the common recommendation](https://golang.org/doc/effective_go#named-results).

And yes, named returned is bad in case of returning two values of different types.

vtopc replied:
By the way, our [Pantheon Go Style Guide](https://github.com/pantheon-systems/go-style-guide/blob/master/style.md) is referencing [Effective Go](https://golang.org/doc/effective_go.html)

vtopc replied:
OK. So no named return, just a question about struct.

zealws replied:
I'm gonna leave this as-is. If you strongly disagree please un-resolve and we can discuss.

vtopc commented on authorization.go:
```diff
@@ -0,0 +1,204 @@
+	words := strings.SplitN(clientCN, ".", 3)
+
+	if len(words) != 3 {
+		return "", "", fmt.Errorf("Unexpected CN format: %q", clientCN)
```
The error in Go should start from the lower case if it's not an abbreviation.

vtopc commented on authorization.go:
```diff
@@ -0,0 +1,204 @@
+// is being denied.
+// See the methods for a description of which Allow* method is chosen depending on the
+// request.
+type AuthorizationCheck interface {
```
nit: what about `-er` suffix https://golang.org/doc/effective_go#interface-names?
e.g.
```suggestion
type AuthorizationChecker interface {
```

zealws replied:
I feel like `Check` is perfectly fine as-is, per it's definition:
>  n. A standard for inspecting or evaluating; a test.

But I don't care that much to argue, so I'll change it.

vtopc commented on authorization.go:
```diff
@@ -0,0 +1,204 @@
+	CNs []string
+}
+
+func (allow AllowSpecificOUandCNs) IsAuthorized(
+	clientOU []string, clientCN string,
+) error {
```
Is it gofmt/goimports compliant? https://github.com/pantheon-systems/go-style-guide/blob/master/style.md#introduction

winmillwill replied:
Why newline at all? I think 80 columns is too small for go, but 100 works great for me, and allows 3 vertical splits on one of my displays.

zealws replied:
> Is it gofmt/goimports compliant

Yes. All my code is:
```
autocmd BufWritePre *.go execute ':GoFmt'
```

90 columns is exactly half my window at 4k with a font size that's readable. That's why I wrapped several of these function definitions since they were just over my window width and the editor would wrap them anyway in a way that's much harder to read.

I'm not gonna argue that everyone should use 90 cause just cause that's what's convenient for me, but I'm gonna stick with that for my code so long as `gofmt` it. :)

vtopc replied:
Good, just unused 😄

vtopc commented on authorization_test.go:
```diff
@@ -0,0 +1,119 @@
+			fallthrough
+		case actualErr != nil && tc.ExpectedErr == nil:
+			fallthrough
+		case actualErr != nil && tc.ExpectedErr != nil &&
```
Just remove `actualErr != nil && tc.ExpectedErr != nil` on this line and you can get rid of all these `fallthrough`

vtopc replied:
oh, sorry and one more case would be needed on top:
```
case actualErr == nil && tc.ExpectedErr == nil:
	break
```

vtopc replied:
meh

zealws replied:
Probably this would be more clear if I wrote an `ErrIsEqual` helper that did all this junk to make sure they're both not null and then checking `.Error()` to ensure they match. But all this is required, simply checking that they're both nil isn't sufficient.

vtopc replied:
Yea. Don't know if need a helper for one-time usage. Let's keep it as it is.

vtopc commented on certauth.go:
```diff
@@ -109,11 +113,20 @@ func (a *Auth) RouterHandler(h httprouter.Handle) httprouter.Handle {
-		h(w, r, ps)
+		ctx := r.Context()
+		if len(a.opt.AllowedOUs) > 0 {
+			ctx = context.WithValue(ctx, HasAuthorizedOU, r.TLS.VerifiedChains[0][0].Subject.OrganizationalUnit)
```
Can `r.TLS.VerifiedChains` be empty?

zealws replied:
No.

`a.Process(...)` called on line 116
calls `a.ValidateRequest` on line 153
which checks that `VerifiedChains` isn't empty and contains valid content on lines 136 and 143

vtopc replied:
Good.

vtopc commented on certauth.go:
```diff
@@ -158,33 +171,45 @@ func (a *Auth) Process(w http.ResponseWriter, r *http.Request) error {
+			if err := v.IsAuthorized(ou, cn); err != nil {
+				return err
+			}
+		} else {
```
Please, use `continue` instead of else - https://github.com/pantheon-systems/go-style-guide/blob/master/style.md#reduce-scope-of-variables

zealws replied:
This function would look like this:

```		if site == "" {
			if err := v.IsAuthorized(ou, cn); err != nil {
				return err
			}
			continue
		}
		if err := v.IsAuthorizedForSite(ou, cn, site); err != nil {
			return err
		}
```

I don't really like the way that reads. `IsAuthorized` and `IsAuthorizedForSite` are both peer functions that are intended be called one or the other, but never both. If they're at the same indentation level that feels better to me.

I think a switch would make the code read a bit better. I'll do this:

```
		switch site {
		case "":
			if err := v.IsAuthorized(ou, cn); err != nil {
				return err
			}
		default:
			if err := v.IsAuthorizedForSite(ou, cn, site); err != nil {
				return err
			}
		}
```

vtopc replied:
> I don't really like the way that reads.

That's Go.

Not critical, keep it as it is.

winmillwill commented on certauth.go:
```diff
@@ -46,6 +45,11 @@ type Options struct {
 	AllowedCNs []string
 
+	// Performs Authorization checks
+	// Each check is called requests to validate that the client is authorized to the
+	// resource. See documentation for AuthorizationCheck for details on the checks.
```
```suggestion
	// Each check validates that the client is authorized to the resource. See documentation for
	// AuthorizationCheck for details on the checks.
```

winmillwill commented on certauth.go:
```diff
@@ -158,33 +171,45 @@ func (a *Auth) Process(w http.ResponseWriter, r *http.Request) error {
-// ValidateCN checks the CN of a verified peer cert and raises a 403 if the CN doesn't match any CN in the AllowedCNs list.
-func (a *Auth) ValidateCN(verifiedCert *x509.Certificate) error {
-	var failed []string
+// ValidateAuth runs each of the AuthorizationChecks configured for the server
```
```suggestion
// CheckAuthorization runs each of the AuthorizationChecks configured for the server
```

winmillwill commented on authorization.go:
```diff
@@ -0,0 +1,204 @@
+	return nil
+}
+
+func (allow AllowSpecificOUandCNs) IsAuthorizedForSite(
```
prefix `Is` kind of implies return of `bool` ... How about `AuthorizeForSite` and `allow.Authorize`?

zealws replied:
> prefix Is kind of implies return of bool

That's fair. I think I'll rename these methods `CheckAuthorization` and `CheckAuthorizationForSite`. I like that better than `Authorize...` cause the latter feels like it has side-effects when it really doesn't.

winmillwill commented on certauth.go:
```diff
@@ -136,7 +149,7 @@ func (a *Auth) ValidateRequest(r *http.Request) error {
 
 // Process is the main Entrypoint
-func (a *Auth) Process(w http.ResponseWriter, r *http.Request) error {
+func (a *Auth) Process(w http.ResponseWriter, r *http.Request, site string) error {
```
What if the `AuthorizationChecker` can just take what it needs from the `http.Request`? Then I think we won't need to change this signature. Otherwise we should v2 the module, IIUC.

zealws replied:
Oh, I had kinda assume most modules did what `go-demo-service` does and don't call this `Process` method directly...

I'll do a cursory check to make sure that nothing is directly calling `Process`. If stuff is, I'll restructure this to not break things, but if there are no callsites outside this repo, I think this is fine as-is?

winmillwill replied:
Proving that there are no callsites outside the repo sounds harder than making a v2 where this *not* exported if that's our expectation. That way we can't break anyone who isn't explicitly opting in, and therefore open to changing their code.

vtopc replied:
That's a public method in the public repo.

vtopc replied:
> What if the AuthorizationChecker can just take what it needs from the http.Request? 

Do you mean header or context?

winmillwill replied:
> That's a public method in the public repo.

So? We're allowed to break whatever we want if we do a `v2`. If we want to keep the `Process` function exported, that's fine too, but if we want to generally be able to change it because we *don't* want people calling it, then we should stop exporting it.

> > What if the AuthorizationChecker can just take what it needs from the http.Request?
> 
> Do you mean header or context?

I mean the whole request. I could see wanting *not* to pass the whole request from a perspective of limiting the degree to which the consuming code has to trust this codebase, but just headers or just context requires something else to stuff the TLS termination information into one or both of those places.

vtopc replied:
@winmillwill, right, I'm just confirming your words.

zealws replied:
> What if the AuthorizationChecker can just take what it needs from the http.Request?

The `AuthorizationChecker` doesn't have access to the `http.Request` right now. Only the site ID derived from the request's URI parameters.

I could make `Process` pull the site ID out instead of doing it in `RouterHandler` but it seems not ideal to parse the arguments twice.

Instead I'm going to restructure `Process` so it keeps it's signature and add a `ProcessWithSite` which accepts the new parameter. This will keep current call sites compiling as is with the current behavior and also allow us to parse the site-id exactly once.

winmillwill reviewed (changes requested):
These are all suggestions/questions. I'm just ticking "Request changes" to ensure we have the conversation.

vtopc commented on authorization.go:
```diff
@@ -0,0 +1,204 @@
+// Returns (site, environment, nil) if the clientCN is valid.
+// Returns ("", "", err) if an error occurs.
+func ParseSiteEnvFromCN(clientCN string) (string, string, error) {
+	// Site CN in format of env.site_uuid.domain
```
Is that some kind of RFC or 3rd party requirements?

winmillwill replied:
it's just what our certs look like

vtopc replied:
Got it.

zealws committed:
Refactored Authorization into an interface

zealws commented:
I think I addressed all the comments thus far. I'm still going to fix add unit-tests so this is still a draft for now.

zealws commented:
Ok, I restructured this again to use a proper interface both for Authorization but also for producing context parameters out of the AuthorizationCheck output.

I also added a new unit-test which tests the httprouter flow. cc @joemiller (since you owned the TODO to do that :wink:  )

Finally, I split out the pantheon-specific stuff to it's own module `github.com/pantheon-systems/go-certauth/pantheon`. I'll add code there tomorrow to do the site auth checks that I had added previously and add unit-tests for that code.

zealws commented:
I changed my mind, this PR is big enough as-is. I'll add the pantheon-specific stuff in a follow-up PR. This is ready to review now.

winmillwill commented on certauth.go:
```diff
@@ -135,56 +182,35 @@ func (a *Auth) ValidateRequest(r *http.Request) error {
+	for _, ck := range a.opt.AuthorizationCheckers {
+		if ps == nil { // not using httprouter
+			parms, err = ck.CheckAuthorization(ou, cn)
+		} else { // using httprouter
```
prefer continue over else

zealws replied:
In this case that's not easy to do because `parms` is set here (or the else at 204) and then read again at 211.

And we want to make sure we only call one of `CheckAuthorization` or `CheckAuthorizationWithParams`, not both.

winmillwill commented on certauth.go:
```diff
@@ -135,56 +182,35 @@ func (a *Auth) ValidateRequest(r *http.Request) error {
+
+	ctxParams := make(map[ContextKey]ContextValue)
+	var (
+		parms map[ContextKey]ContextValue
```
params?

zealws replied:
I can change it if my short variable names bug you :stuck_out_tongue:

winmillwill commented on certauth.go:
```diff
@@ -109,15 +114,57 @@ func (a *Auth) RouterHandler(h httprouter.Handle) httprouter.Handle {
+// Returns an http.Request with additional context values applied, or an error if
+// something went wrong.
+func (a *Auth) ProcessWithParams(
+	w http.ResponseWriter, r *http.Request, ps httprouter.Params,
```
The httprouter.Params come from the request context anyway, FWIW.

zealws replied:
This is true, but they are already parsed by `httprouter` and I wanted to avoid the extra effort since it's not required.

winmillwill commented on authorization_test.go:
```diff
@@ -0,0 +1,105 @@
+	vl := v.([]string)
+	if !ok {
+		t.Errorf("Expected context key %q was not present %v", HasAuthorizedOU.String(), params)
+	} else if len(vl) != 1 || vl[0] != actualOU {
```
drop the else or `t.FailNow` above?

zealws replied:
same as above (although in this case it's the end of the method so it doesn't matter as much

winmillwill commented on authorization_test.go:
```diff
@@ -0,0 +1,105 @@
+	v, ok := params[HasAuthorizedCN]
+	if !ok {
+		t.Errorf("Expected context key %q was not present %v", HasAuthorizedCN.String(), params)
+	} else if v != actualCN {
```
Drop the else or `t.FailNow` above?

zealws replied:
`t.FailNow` prevents the rest of this test method from running so I tend to avoid it unless that specific assertion makes the rest of the test invalid for some reason, which is definitely not the case here (it only makes the following check invalid, not the whole method).

This would be less of a problem if the tests were smaller, but the pattern seemed to be writing larger tests with several cases for a specific feature or method, rather than small targeted micro tests. I might just restructure all the tests to use that format as I generally prefer smaller more targeted tests.

zealws replied:
I'm gonna file a follow-up PR with a bunch of changes to these tests.

winmillwill commented on authorization_test.go:
```diff
@@ -0,0 +1,105 @@
+		_, err := check.CheckAuthorization([]string{""}, tc.ActualCN)
+
+		if err != nil && tc.IsAllowed {
+			t.Errorf("Expected AllowedCNs (%v) and ActualCN (%v) to pass validation, but it failed: err: %s", tc.AllowedCNs, tc.ActualCN, err)
```
Having difficulty figuring out when you are and are not ok with long lines. Seems strange to newline function parameters and leave this as is.

zealws replied:
I wrap these by hand cause auto-wrap in VIM tends to get undone by `go fmt`. Occasionally I miss a line :shrug:
